Update KEP-5007 DRA Device Binding Conditions (#5342)
kubernetes/enhancements: keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

To address this, we propose a mechanism — `BindingConditions` — that allows the scheduler to defer Pod binding until the required device is confirmed to be ready.
This improves reliability by avoiding premature binding and enables better coordination with external device controllers.

While the original motivation came from fabric-attached devices, the mechanism is designed to be broadly applicable.
It can support other scenarios where resource readiness is asynchronous or failure-prone, such as remote accelerators or gang scheduling.
This proposal focuses on enabling readiness-aware binding as a general scheduling enhancement.

### Goals

1. **Enable Readiness-Aware Binding**:  
   Introduce a mechanism (`BindingConditions`) that allows the scheduler to defer Pod binding until external resources are confirmed to be ready.
   This improves scheduling reliability and avoids premature binding in environments where resource preparation is asynchronous or failure-prone.

2. **Define Binding and Failure Conditions Explicitly**:  
   Allow device providers to specify `BindingConditions` and `BindingFailureConditions` that the scheduler can evaluate during the PreBind phase.

> 3. **Prioritize Devices Based on Readiness**:  

Reviewer: Ack, having none vs something seems good enough heuristic for the beginning.

> 3. **Prioritize Devices Based on Readiness**:  

Author: Thanks for the feedback.
Yes, the current logic prioritizes devices without BindingConditions. This is because devices without conditions are immediately usable and don't require waiting in the PreBind phase.
This prioritization is already implemented in the GatherPools() function on my PR.

### Non-Goals

- Defining a complete model for fabric-attached device coordination.
- While this KEP introduces a mechanism that supports such use cases, broader architectural questions — such as how to model attachment workflows or coordinate between node-local and fabric-aware components — will be addressed in follow-up discussions.

## Proposal

This proposal introduces a mechanism that allows the scheduler to defer Pod binding until external resources are confirmed to be ready.
This is achieved by extending the `ResourceSlice` and `DeviceRequestAllocationResult` APIs with the following fields:

- `BindingConditions`: a list of condition keys that must be satisfied (i.e., set to `True`) before binding can proceed.
- `BindingFailureConditions`: a list of condition keys that, if any are `True`, indicate that binding should be aborted and the Pod rescheduled.
- `BindingTimeoutSeconds`: a timeout value (in seconds) that defines how long the scheduler should wait for all `BindingConditions` to be satisfied. If the timeout is exceeded, the allocation is cleared and the Pod is rescheduled.

> - `UsageRestrictedToNode`: a boolean field (optional) that indicates whether the scheduler should set the nodename to the ResourceClaim's nodeSelector during the scheduling process. When set to true, the scheduler records the selected node name in the claim, effectively reserving the resource for that specific node. This is particularly useful for external controllers during the PreBind phase, as they can retrieve the node name from the claim and perform node-specific operations such as device attachment or preparation.

Reviewer 2: Are you suggesting that the controller which needs to set binding conditions should look up the pod through its `ReservedFor` entry and then from the pod figure out which node is being targeted? That doesn't work because the pod hasn't actually been bound to that node yet.

> I also expect that some devices that needs binding conditions, does not have to be restricted to a single node, since they could be multi-node devices. 

What is restricted to a single node is the ResourceClaim, not the device. I think that is consistent with the purpose of the KEP, which is about ensuring that a pod can run on one specific node. Theoretically it might be possible to attach the same device to a set of nodes and then avoid updating the ResourceClaim when the pod needs to be scheduled to a different node, but that seems unlikely to matter much in practice.

If we wanted to be more liberal than `UsageRestrictedToNode`, then we would need a different way to describe the set of nodes after binding. We cannot assume that it is the same set as before binding, and the set might depend on the target node - this seems very complex and isn't supported by some actual use case.

> - `UsageRestrictedToNode`: a boolean field (optional) that indicates whether the scheduler should set the nodename to the ResourceClaim's nodeSelector during the scheduling process. When set to true, the scheduler records the selected node name in the claim, effectively reserving the resource for that specific node. This is particularly useful for external controllers during the PreBind phase, as they can retrieve the node name from the claim and perform node-specific operations such as device attachment or preparation.

Author: @dom4ha, thank you for your comment. I'd like to add a few thoughts to clarify the reasoning behind `UsageRestrictedToNode`.

While `ReservedFor` can reference the Pod, as @pohly mentioned, the Pod has not yet been bound to a node during the PreBind phase. This means we cannot reliably determine the target node from the Pod object at that point.

Since `BindingConditions` are evaluated during PreBind, external controllers that need to prepare node-specific resources (e.g., device attachment) must know which node the Pod is expected to run on. This is where `UsageRestrictedToNode: true` becomes essential. It allows the scheduler to write the selected node name into the `NodeSelector` of the `ResourceClaim` before PreBind. This enables the controller to determine the target node and perform the necessary setup accordingly.

Regarding the concern that devices requiring BindingConditions don't necessarily need to be restricted to a single node: that's a valid point. However, what UsageRestrictedToNode restricts is not the device or the ResourceSlice itself, but rather whether the ResourceClaim is limited to a specific node. This aligns with the KEP’s goal of ensuring that a Pod can reliably run on one specific node.

I hope this helps clarify the motivation behind `UsageRestrictedToNode`.

> - `UsageRestrictedToNode`: a boolean field (optional) that indicates whether the scheduler should set the nodename to the ResourceClaim's nodeSelector during the scheduling process. When set to true, the scheduler records the selected node name in the claim, effectively reserving the resource for that specific node. This is particularly useful for external controllers during the PreBind phase, as they can retrieve the node name from the claim and perform node-specific operations such as device attachment or preparation.

Reviewer: > This is where UsageRestrictedToNode: true becomes essential. It allows the scheduler to write the selected node name into the NodeSelector of the ResourceClaim before PreBind. 

In which phase do you think the `NodeSelector` should be set then?

Note that pre-bind is currently the first one that is performed asynchronously with regard to the scheduling loop, so any api-calls that would be part of the earlier phases may by blocking scheduling. I however think it should be still fine to set `NodeSelector` at the beginning of pre-bind.

> While ReservedFor can reference the Pod, as [Patrick Ohly](https://github.com/pohly) mentioned, the Pod has not yet been bound to a node during the PreBind phase. This means we cannot reliably determine the target node from the Pod object at that point.

Ah right. We want to put that information to the Pod during pre-bind (set `nominatedNodeName`) in KEP https://github.com/kubernetes/enhancements/pull/5287, but this field may be more ambiguous, so we should think twice before using it. Independently from it, we work on reservations concept which should be more suitable for determining pods assignment (before binding), but the problem is that it's not there yet, so we cannot use it.

@wojtek-t @sanposhiho @macsko 

> However, what UsageRestrictedToNode restricts is not the device or the ResourceSlice itself, but rather whether the ResourceClaim is limited to a specific node.

I can imagine a ResourceClaim that is shared between many pods and it attaches a devices to a host machine on which several nodes can run. In that scenario the first pod triggers attachment tho the correct host based on pod future assignment. In such case, the `NodeSelector` should specify all nodes the following pods can be bound to, but not only the first pod's one.

Summarizing, I think it's fine to use `NodeSelector` in alpha, but I'd think about the alternatives for beta (we already plan to rethink the approach for handling attachable devices). Expressing the target node by `NodeSelector` sounds to me a bit hacky approach, which is filling a gap where we don't have more explicit way of expressing intended binding/attachment.

> - `UsageRestrictedToNode`: a boolean field (optional) that indicates whether the scheduler should set the nodename to the ResourceClaim's nodeSelector during the scheduling process. When set to true, the scheduler records the selected node name in the claim, effectively reserving the resource for that specific node. This is particularly useful for external controllers during the PreBind phase, as they can retrieve the node name from the claim and perform node-specific operations such as device attachment or preparation.

Reviewer 2: > In which phase do you think the NodeSelector should be set then?

When the scheduler updates the claim with the allocation result, which then triggers the controller's binding logic. Every allocation result has such a NodeSelector which is determined based on what was allocated.

> I can imagine a ResourceClaim that is shared between many pods and it attaches a devices to a host machine on which several nodes can run. In that scenario the first pod triggers attachment tho the correct host based on pod future assignment. In such case, the NodeSelector should specify all nodes the following pods can be bound to, but not only the first pod's one.

If it literally attaches "a devices to a (= one!) host machine" (emphasis mine), then a node selector which matches exactly that single node would be correct. I think you want to support a "binding device to a set of nodes" use case instead of the current "bind device to a single node". That's worth keeping in mind for later, but right now it's out of scope.

So with this clarified, okay to approve?

> - `UsageRestrictedToNode`: a boolean field (optional) that indicates whether the scheduler should set the nodename to the ResourceClaim's nodeSelector during the scheduling process. When set to true, the scheduler records the selected node name in the claim, effectively reserving the resource for that specific node. This is particularly useful for external controllers during the PreBind phase, as they can retrieve the node name from the claim and perform node-specific operations such as device attachment or preparation.

Reviewer 2: > do I understand correctly that no new api call would be needed and setting NodeSelector would be effectively in pre-bind?

Yes.

> The KEP wording suggest it happens somewhere before pre-bind, thus my confusion.

We should fix that. Can you point to the line where this isn't clear?

> Do I understand correctly that other nodes would not see this device and pods scheduled there won't be able to use it? I guess it doesn't matter if a gpu device is not shareable and supporting sharing is out of scope.

Yes, that's exactly it. The device isn't bound to the physical machine hosting several nodes, it gets bound to one node. As you said, lacking some more complex sharing support that's exactly what is meant to be supported.

)
```

> #### DeviceRequestAllocationResult Enhancements

Author: Yes, that's right, I'll update it to include a description about `AllocationTimestamp`.

When `UsageRestrictedToNode: true` is set, the scheduler DRA plugin will perform the following steps:

> 1. **Set NodeSelector in api-server**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.

Reviewer: ```suggestion
1. **Set NodeSelector in api-server**: During the `PreBind` phase, propagate the `NodeName` to the `ResourceClaim`'s `NodeSelector` to the api-server.
```

> If Conditions are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:

Reviewer: ```suggestion
If Conditions are present, the scheduler DRA plugin will additionally perform the following steps:
```

If Conditions are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:

> 

Reviewer: 1. **Set NodeSelector in api-server**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.

Reviewer: I have still some small corrections

-->

> - `<package>`: `<date>` - `<test coverage>`

Reviewer 3: Remove this line

> - <test>: <link to test coverage>

Reviewer 3: Remove this line

- when the KEP was retired or superseded
-->

- 2024-12: Initial proposal drafted and discussed in SIG scheduling, WG device management.
- 2025-03: KEP revised to introduce BindingConditions as a general mechanism.

> - 2025-05: Updated KEP submitted for review.

Reviewer 3: Please shortly describe the update
